lib/active_merchant/billing/gateways/versa_pay.rb

```diff
@@ -0,0 +1,307 @@
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def success_from(response)
+        response['success'] || false
+      end
+
+      def message_from(response)
+        first_transaction = response['transactions']&.first
+        gateway_response_errors = gateway_errors_message(response)
+
+        response_message = {
+          errors: response['errors']&.join(', ').presence,
+          gateway_error_message: first_transaction&.dig('gateway_error_message').presence,
+          gateway_response_errors: gateway_response_errors.presence
+        }.compact
+
+        success_from(response) ? 'Succeeded' : response_message.map { |key, value| "#{key}: #{value}" }.join(' | ')
+      end
+
+      def authorization_from(response)
+        response['transaction']
+      end
+
+      def error_code_from(response)
+        return if success_from(response)
+
+        first_transaction = response['transactions']&.first
+        error_info = {
+          gateway_error_code: first_transaction&.dig('gateway_error_code'),
+          response_code: response['response_code']
+        }.compact
+
+        error_info.map { |key, value| "#{key}: #{value}" }.join(' | ')
+      end
+
+      def gateway_errors_message(response)
+        errors = response.dig('transactions', 0, 'gateway_response', 'errors')
+        return unless errors.is_a?(Hash)
+
+        errors.flat_map do |field, error_details|
+          error_details.flat_map do |error|
+            if error.is_a?(Hash)
+              error.map { |key, messages| "[#{field} - #{key}: #{messages.join(', ')}]" }
+            else
+              "[#{field} - #{error}]"
+            end
+          end
+        end.join(' , ')
+      end
+
+      def url(endpoint)
+        "#{test? ? test_url : live_url}/api/gateway/v1/orders/#{endpoint}"
+      end
+
+      def basic_auth
+        Base64.strict_encode64("#{@api_token}:#{@api_key}")
+      end
+
+      def request_headers
+        {
+          'Content-Type' => 'application/json',
+          'Authorization' => "Basic #{basic_auth}"
+        }
+      end
+
+      def dig_cvv_code(first_transaction)
+        return unless first_transaction
+
+        first_transaction.dig('cvv_response') ||
+          first_transaction.dig('gateway_response', 'cvv_response') ||
+          first_transaction.dig(
+            'gateway_response',
+            'gateway_response',
+            'response', 'content',
+            'create',
+            3,
+            'transaction',
+            'cvvresponse'
+          )
+      end
+
+      def dig_avs_code(first_transaction)
+        return unless first_transaction
+
+        first_transaction.dig('avs_response') ||
+          first_transaction.dig('gateway_response', 'avs_response') ||
+          first_transaction.dig(
+            'gateway_response',
+            'gateway_response',
+            'response', 'content',
+            'create',
+            3,
```

Comment:
💬 Comment:

Versapay really really likes to complicate things, I can see why all this dig structure, one suggestion:

I'm not sure if we can ensure that avs_responde element is always going to be the 4th one, so maybe digging down to create and then an small `find` to check if an element has the `avsresponse`